src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,316 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract partial class AbstractTypeImportCompletionProvider : CommonCompletionProvider
+    {
+        private readonly SyntaxAnnotation _annotation = new SyntaxAnnotation();
+
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract ImmutableHashSet<string> GetNamespacesInScope(SyntaxNode location, SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        private readonly ITypeImportCompletionService _typeImportCompletionService;
+        private readonly IExperimentationService _experimentationService;
+
+        public AbstractTypeImportCompletionProvider(Workspace workspace)
+            : base()
+        {
+            _typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+            _experimentationService = workspace.Services.GetService<IExperimentationService>();
```

Comment:
it seems unnecessary to do this.  Don't you have the Document when completion runs?  you can get the Workspace from that.

Reply:
Any benefits of doing that instead of saving them as fields?

Reply:
it's just fits the model we have today where you get what you need from what is passed in, instead of needing another way to get at this data.  none of the other compleiton providers work this way, so i'd prefer we just stick with the existing model.

Reply:
ok, fixed